google/cloud/pubsub/subscriber.h

```diff
@@ -116,11 +114,9 @@ class Subscriber {
    * https://en.cppreference.com/w/cpp/utility/functional/function
    */
   template <typename Callable>
-  future<Status> Subscribe(Subscription const& subscription, Callable&& cb,
-                           SubscriptionOptions options = {}) {
+  future<Status> Subscribe(Callable&& cb) {
     std::function<void(Message, AckHandler)> f(std::forward<Callable>(cb));
-    return connection_->Subscribe(
-        {subscription.FullName(), std::move(f), std::move(options)});
+    return connection_->Subscribe({std::move(f)});
```

Comment:
Is the additional SubscriberConnection abstraction that useful if Subscriber is just going to delegate the calls to it? I wonder what benefit we are gaining by having these two classes be separate. We don't have to make the change in this PR, but maybe it's worth us considering as a future change.

Reply:
In the C++ libraries all `*Client` classes (which `Subscriber` models) are value types.  But we need to support mocking, so the `*Connection` classes are pure abstract.

Reply:
Is there a reason it is done this way as opposed to making a mockable `*Client`? It seems like an extra level of indirection where one has to understand the implementation of the `*Client` class in order to see which calls are made to the underlying `*Connection` class. If I'm implementing a publisher or subscriber, I don't want to have to look into the implementation of the `Publisher` or `Subscriber` class in order to write unit tests.

Reply:
> Is there a reason it is done this way as opposed to making a mockable `*Client`?

Mockable classes are not value types.  Value types are easier to use in C++.

> It seems like an extra level of indirection where one has to understand the implementation of the `*Client` class in order to see which calls are made to the underlying `*Connection` class.

Not to *use* the `*Client` class, right?

> If I'm implementing a publisher or subscriber, I don't want to have to look into the implementation of the `Publisher` or `Subscriber` class in order to write unit tests.

Fair enough, that is the tradeoff we made: simpler to use `*Client` objects, for a cost in difficulty to write unit tests.

In any case, we are not going to change this for Pub/Sub: consistency across C++ libraries is also important.